### PR TITLE
Make runtime tests run with Android x64

### LINKED ISSUE #41414: Android Runtime Test Design Proposal
This is a proposal for how to get the Dotnet runtime tests running for Android. Not everyhting is pinned down yet, and some things need to be investigated.

### Existing Test Infrastructure

Once built, the runtime test have three parts: 

1. An xunit wrapper
2. A shell script
3. The actual test assembly.

The XUnit wrapper runs serves as an interface between the shell script and Xunit. It calls the shell script. The shell script sets up various environment variables, and calls the test assembly. The test assembly itself is just a program that exists with 100 if the test passes.

### What needs to change for android

The android platform has limitations that mean we can't just change the OS to "android" and expect the runtime tests to work. Among the most salient: We cannot fork a process on android, so there is no way for the xunit wrapper to start a shell script, or for the shell script to run the test assembly.

 Instead, we will need to create an *Android Test Runner*. The Android Test Runner is an android application, that when started with a test name, runs that test and returns the result.  The shell script will run on the host OS, and instead of directly running the test assembly, will invoke the Android Test Runner with the name of a test as a parameter.

### Build Changes

The android test runner must be built as part of the building the tests. It needs to contain all the test assemblies inside it.

Android simulator should probably be started as a helix precommand.
     
### Helix payload changes

When running in CI, the Xunit wrappers (and thus the scripts), are all run on Helix. Thus the built Android Test Runner must be included in te Helix payload.

### Unknowns

 Is there a practical limit on how big an android app can be? Would including ~3k test assemblies exceed this limit?

 Would an alternative design where we build one android app per test be too slow?

@fanyang-mono  @imhameed @SamMonoRT

### PR DESCRIPTION
**Pull request by fanyang-mono**
This PR is to enable runtime tests running with Android. The design of this work was captured by https://github.com/dotnet/runtime/issues/41414.

This PR filtered out 51 tests from running with Android in issues.target file. Additionally, it stopped 3 test projects from building for Android, due to the incompatibility of the tests. The reasons have been captured as comments in each file. In total, 2293 tests were running and passed on the newly added CI lane.

Fixes #41414

### CODE REVIEW

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+parameters:
```
Comment (mdh1418) [inline eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml:1]: The spacing throughout is off compared to the other `.yml` files in the same folder.

File: src/tests/run.sh [hunk 2]
```diff
@@ -332,6 +336,9 @@ if [ $buildArch = "wasm" ]; then
     runtestPyArguments+=("-os" "Browser")
 fi
 
+if [ $buildOS = "Android" ]; then
+    runtestPyArguments+=("-os" "Android")
```
Comment (mdh1418) [inline src/tests/run.sh:340]: Why is this needed now? Did this not work before when the script is supplied with `-os Android` before or was that never enabled?

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml [hunk 3]
```diff
@@ -140,4 +140,8 @@ jobs:
     - ${{ if eq(parameters.platform, 'Browser_wasm') }}:
       - Ubuntu.1804.Amd64.Open
 
+    # Android
```
Comment (mdh1418) [inline eng/pipelines/coreclr/templates/helix-queues-setup.yml:143]: It feels like the jobs should be in alphabetical order. If its possible, could we move this and WebAssembly (Maybe change to Browser wasm to follow the convention of os arch?) 
Also conventionally it looks like the jobs are split by arch, so split into Android x86 and Android x64?

File: src/tests/run.sh (hunk 2 above)
Comment (fanyang-mono) [reply src/tests/run.sh:340]: So for runtime tests, from terminal, you invoke `run.sh`, then `run.sh` invoke `runtest.py` to do more work. This is to pass OS configuration that you specified to `run.sh` down to `runtest.py`.

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  
+          publishTestResults: true
+  
+          timeoutPerTestInMinutes: $(timeoutPerTestInMinutes)
+          timeoutPerTestCollectionInMinutes: $(timeoutPerTestCollectionInMinutes)
+  
+          runCrossGen: ${{ and(eq(parameters.readyToRun, true), ne(parameters.crossgen2, true)) }}
+          runCrossGen2: ${{ and(eq(parameters.readyToRun, true), eq(parameters.crossgen2, true)) }}
+          compositeBuildMode: ${{ parameters.compositeBuildMode }}
+          runInUnloadableContext: ${{ parameters.runInUnloadableContext }}
```
Comment (trylek) [inline eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml:74]: I believe that none of these four arguments (runCrossGen, runCrossGen2, compositeBuildMode and runInUnloadableContext) applies to Android tests.

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  
+        ${{ if eq(variables['System.TeamProject'], 'public') }}:
+          creator: $(Build.DefinitionName)
+  
+          helixBuild: $(Build.BuildNumber)
+          helixSource: $(_HelixSource)
+  
+          # REVIEW: not sure why "cli" is part of the names here. Leave it for the ones that already had it,
+          # but don't add it to new ones.
+          ${{ if eq(parameters.readyToRun, true) }}:
```
Comment (trylek) [inline eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml:55]: Not applicable to Android tests. (Nor to WASM, for that matter.)

File: src/coreclr/tests/helixpublishwitharcade.proj [hunk 6]
```diff
@@ -82,6 +82,7 @@
     <HelixRuntimeRid Condition="'$(TargetOSSpec)' == 'Linux'">linux-$(TargetArchitecture)</HelixRuntimeRid>
     <HelixRuntimeRid Condition="'$(TargetOSSpec)' == 'Linux_musl'">linux-musl-$(TargetArchitecture)</HelixRuntimeRid>
     <HelixRuntimeRid Condition="'$(TargetArchitecture)' == 'wasm'">browser-wasm</HelixRuntimeRid>
+    <HelixRuntimeRid Condition="'$(TargetOS)' == 'Android'">android-x64</HelixRuntimeRid>
```
Comment (mdh1418) [inline src/coreclr/tests/helixpublishwitharcade.proj:85]: It looks like `TargetOSSubgroup` is empty, could we also use `TargetOSSpec` here?
`<HelixRuntimeRid Condition="'$(TargetOSSpec)' == 'Android'">android-$(TargetArchitecture)</HelixRuntimeRid>`

I'm wondering if we can change 
`<HelixRuntimeRid Condition="'$(TargetArchitecture)' == 'wasm'">browser-wasm</HelixRuntimeRid>`
to be 
`<HelixRuntimeRid Condition="'$(TargetOSSpec)' == 'Browser'">browser-$(TargetArchitecture)</HelixRuntimeRid>`

File: src/tests/Common/CLRTest.Execute.Bash.targets [hunk 7]
```diff
@@ -304,6 +319,20 @@ $(BashLinkerTestCleanupCmds)
...
+      <![CDATA[
+# run Android app
+AdbTool=$ANDROID_SDK_ROOT/platform-tools/adb
+$AdbTool kill-server
+$AdbTool start-server
+$AdbTool logcat -c" ContinueOnError="WarnAndContinue
+$AdbTool uninstall net.dot.$(MsBuildProjectName)" ContinueOnError="WarnAndContinue
+$AdbTool install apk/bin/$(MsBuildProjectName).apk
+$AdbTool shell am instrument -w net.dot.$(MsBuildProjectName)/net.dot.MonoRunner
+$AdbTool logcat -d -s DOTNET
```
Comment (EgorBo) [inline src/tests/Common/CLRTest.Execute.Bash.targets:335]: It should use xharness instead, I don't actually remember why the Android sample still uses raw `adb`

File: tools-local/tasks/mobile.tasks/AndroidAppBuilder/Templates/monodroid.c [hunk 8]
```diff
@@ -138,9 +138,9 @@ mono_mobile_runtime_init (void)
 {
     // uncomment for debug output:
     //
-    // setenv ("XUNIT_VERBOSE", "true", true);
-    // setenv ("MONO_LOG_LEVEL", "debug", true);
-    // setenv ("MONO_LOG_MASK", "all", true);
+    setenv ("XUNIT_VERBOSE", "true", true);
+    setenv ("MONO_LOG_LEVEL", "debug", true);
+    setenv ("MONO_LOG_MASK", "all", true);
```
Comment (EgorBo) [inline tools-local/tasks/mobile.tasks/AndroidAppBuilder/Templates/monodroid.c:143]: it'll ddos CI 😄

File: src/coreclr/tests/tests.targets [hunk 9]
```diff
@@ -54,7 +54,7 @@
 
       <CorerunExecutable Condition="'$(RunningOnUnix)' == 'true'"  >$(CORE_ROOT)\corerun</CorerunExecutable>
       <CorerunExecutable Condition="'$(RunningOnUnix)' != 'true'">$(CORE_ROOT)\corerun.exe</CorerunExecutable>
-      <CorerunExecutable Condition="'$(TargetArchitecture)' == 'wasm'"> $(DotnetRoot)/dotnet</CorerunExecutable>
+      <CorerunExecutable Condition="'$(TargetArchitecture)' == 'wasm' Or '$(TargetOS)' == 'Android'"> $(DotnetRoot)/dotnet</CorerunExecutable>
```
Comment (mdh1418) [inline src/coreclr/tests/tests.targets:57]: I noticed that `CorerunExecutable` is also in `src/tests/run.proj`, is this line not needed there?

File: src/coreclr/tests/helixpublishwitharcade.proj (hunk 6 above)
Comment (trylek) [inline src/coreclr/tests/helixpublishwitharcade.proj:85]: In the helix-queues-setup you seem to be checking for "android-x64" and "android-x86" - should this be somehow forked for the two architectures?

File: src/tests/Common/CLRTest.Execute.Bash.targets [hunk 10]
```diff
@@ -272,7 +272,22 @@ $__Command msbuild $CORE_ROOT/wasm-test-runner/WasmTestRunner.proj /p:NetCoreApp
...
+else
+  __Command+=" dotnet"
+fi
+
+$__Command publish $CORE_ROOT/android-test-runner/AndroidTestRunner.proj -c $(Configuration) -r android-$(TargetArchitecture) /p:Platform=$(TargetArchitecture) /p:TargetArchitecture=$(TargetArchitecture) /p:NetCoreAppCurrent=$(NetCoreAppCurrent) /p:TestAssembly=`pwd`/$(MsBuildProjectName).dll /p:TestBinDir=`pwd`
+
+   ]]>
+      </BashCLRTestPreCommands>
+
+      <BashCLRTestLaunchCmds Condition="'$(CLRTestKind)' == 'BuildAndRun' And $(TargetArchitecture) != 'wasm' And '$(TargetOS)' != 'Android'">
```
Comment (mdh1418) [inline src/tests/Common/CLRTest.Execute.Bash.targets:290]: Isn't ` And $(TargetArchitecture) != 'wasm' And '$(TargetOS)' != 'Android'` redundant since the property is overridden below by 
`<BashCLRTestLaunchCmds Condition="'$(CLRTestKind)' == 'BuildAndRun' And $(TargetArchitecture) == 'wasm' ">`
and 
`<BashCLRTestLaunchCmds Condition="'$(CLRTestKind)' == 'BuildAndRun' And $(TargetOS) == 'Android' ">`

File: src/tests/Common/Directory.Build.targets [hunk 11]
```diff
@@ -129,6 +129,32 @@
           TargetDir="runtime-test/"/>
     </ItemGroup>
 
+    <ItemGroup Condition="'$(TargetOS)' == 'Android'">
+        <RunTimeDependencyCopyLocal
+          Include="@(RuntimeFiles)"
+          TargetDir="runtimepack/native/"/>
+
+        <RunTimeDependencyCopyLocal
```
Comment (mdh1418) [inline src/tests/Common/Directory.Build.targets:137]: These are also needed for Android? The comment for wasm made it seem like a special case.

File: src/tests/Common/dirs.proj [hunk 12]
```diff
@@ -18,9 +18,10 @@
     </ItemGroup>
 
     <ItemGroup>
-      <AllProjects Include="$(TestRoot)**\*.csproj" Exclude="@(DisabledProjects)" />
+      <!-- <AllProjects Include="$(TestRoot)**\*.csproj" Exclude="@(DisabledProjects)" />
       <AllProjects Include="$(TestRoot)**\*.fsproj" Exclude="@(DisabledProjects)" />
-      <AllProjects Include="$(TestRoot)**\*.ilproj" Exclude="@(DisabledProjects)" />
+      <AllProjects Include="$(TestRoot)**\*.ilproj" Exclude="@(DisabledProjects)" /> -->
+      <AllProjects Include="$(TestRoot)JIT\Methodical\divrem\div\i4div_cs_do.csproj" />
```
Comment (mdh1418) [inline src/tests/Common/dirs.proj:24]: Where does this come from?

File: src/tests/Directory.Build.targets [hunk 13]
```diff
@@ -161,7 +161,7 @@
 
      <!-- There are currently no native project binaries on wasm -->
      <Error  Text="The native project files are missing in $(NativeProjectOutputFolder) please run build from the root of the repo at least once"
-             Condition="'@(NativeProjectBinaries)' == '' And '$(TargetArchitecture)' != 'wasm'" />
+             Condition="'@(NativeProjectBinaries)' == '' And '$(TargetArchitecture)' != 'wasm' And '$(TargetOS)' != 'Android'"/>
```
Comment (mdh1418) [inline src/tests/Directory.Build.targets:164]: Could the comment above be updated to include Android if there are no project binaries for android either?

File: src/tests/run.proj [hunk 14]
```diff
@@ -66,9 +66,9 @@
     <Message Text="Found $(TestCount) built tests"/>
 
     <ItemGroup>
-      <Error Condition="'$(CLRTestPriorityToBuild)' == '0' and '$(TestCount)' &lt;= 2000" Text="Unexpected test count. Expected &gt; 2000, found $(TestCount).'" />
+      <!-- <Error Condition="'$(CLRTestPriorityToBuild)' == '0' and '$(TestCount)' &lt;= 2000" Text="Unexpected test count. Expected &gt; 2000, found $(TestCount).'" />
```
Comment (mdh1418) [inline src/tests/run.proj:69]: Why are these commented out?

File: src/coreclr/tests/runtest.py [hunk 15]
```diff
@@ -1002,7 +1002,7 @@ def setup_args(args):
         location using the build type and the arch.
     """
 
-    requires_coreroot = args.arch.lower() != "wasm"
+    requires_coreroot = args.arch.lower() != "wasm" and args.host_os.lower() != "android"
```
Comment (trylek) [inline src/coreclr/tests/runtest.py:1005]: This arbitrary mixing of "arch" vs. "host_os" checks is just asking for trouble - for instance, are we expecting to have "android wasm" tests at some point? I don't have any particular suggestions how to improve this right now, I'm mentioning this more or less as a heads-up that this might be worth clarifying at some point.

File: src/tests/Common/android-test-runner/AndroidTestRunner.proj [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    <EnableTargetingPackDownload>false</EnableTargetingPackDownload>
+  </PropertyGroup>
+
+  <UsingTask TaskName="AndroidAppBuilderTask"
+             AssemblyFile="$(AndroidAppBuilderDir)\AndroidAppBuilder.dll" />
+
+  <!-- Build APK during 'dotnet publish' -->
+  <Target Name="BuildAppBundle" AfterTargets="Build">
+    <PropertyGroup>
+      <AndroidAbi Condition="'$(Platform)'=='arm64'">arm64-v8a</AndroidAbi>
```
Comment (trylek) [inline src/tests/Common/android-test-runner/AndroidTestRunner.proj:15]: What architectures do we intend to support on Android? This bit refers to arm64 / arm / x64, in another place we're checking for "android_x64 & android_x86" and in yet another place we hardcode the RID to be "android_x64". It might be useful to be more explicit about "what android architectures make sense or are planned to be ultimately supported" vs. "what android architectures are currently supported" and ideally formulate most of the script provisions for the general case even though right now you're just standing up a single targeting architecture.

File: src/tests/Common/dirs.proj [hunk 17]
```diff
@@ -18,9 +18,10 @@
     </ItemGroup>
 
     <ItemGroup>
-      <AllProjects Include="$(TestRoot)**\*.csproj" Exclude="@(DisabledProjects)" />
+      <!-- <AllProjects Include="$(TestRoot)**\*.csproj" Exclude="@(DisabledProjects)" />
```
Comment (trylek) [inline src/tests/Common/dirs.proj:21]: I assume this is just local instrumentation not to be merged in, otherwise we need to change this to android-specific test filtering unless you intend to turn off most of CoreCLR testing ;-).

**Review by trylek — COMMENTED** [review]
Mostly looks pretty good, thanks Fan! I'm looking forward to the E2E change.

File: src/tests/run.proj (hunk 14 above)
Comment (trylek) [reply src/tests/run.proj:69]: These are sanity checks to make sure someone doesn't inadvertently turn most of the tests off (just like Fan did in the dirs.proj change, please see my comment there).

File: src/coreclr/tests/tests.targets (hunk 9 above)
Comment (fanyang-mono) [reply src/coreclr/tests/tests.targets:57]: That is for a different target `MonoAotCompileTests`, which is irrelevant here.

File: src/tests/Common/CLRTest.Execute.Bash.targets (hunk 10 above)
Comment (fanyang-mono) [reply src/tests/Common/CLRTest.Execute.Bash.targets:290]: They are mutually exclusive. Those are 3 different scenarios. None of them couldn't be overwritten by any other two scenarios.

File: tools-local/tasks/mobile.tasks/AndroidAppBuilder/Templates/monodroid.c (hunk 8 above)
Comment (fanyang-mono) [reply tools-local/tasks/mobile.tasks/AndroidAppBuilder/Templates/monodroid.c:143]: Will revert this. It is for debugging temporarily.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml (hunk 3 above)
Comment (fanyang-mono) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml:143]: According to https://github.com/dotnet/runtime/issues/33381, only x64 is needed. I will remove x86.

File: src/coreclr/tests/helixpublishwitharcade.proj (hunk 6 above)
Comment (fanyang-mono) [reply src/coreclr/tests/helixpublishwitharcade.proj:85]: According to https://github.com/dotnet/runtime/issues/33381, only x64 is needed. I will remove x86.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml (hunk 3 above)
Comment (fanyang-mono) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml]: Fixed order.

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml (hunk 1 above)
Comment (fanyang-mono) [reply eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml:1]: Fixed format

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml (hunk 4 above)
Comment (fanyang-mono) [reply eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml]: Removed.

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml (hunk 5 above)
Comment (fanyang-mono) [reply eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml]: Removed.

File: src/tests/Common/Directory.Build.targets (hunk 11 above)
Comment (fanyang-mono) [reply src/tests/Common/Directory.Build.targets:137]: Android and wasm fall into the same special case category. So yes, they are needed.

File: src/tests/Common/dirs.proj (hunk 12 above)
Comment (fanyang-mono) [reply src/tests/Common/dirs.proj:24]: This is a hack for faster development. It will be reverted before merged.

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml [hunk 18]
```diff
@@ -1,83 +1,76 @@
...
+
+      ${{ if eq(variables['System.TeamProject'], 'public') }}:
+        creator: $(Build.DefinitionName)
+
+        helixBuild: $(Build.BuildNumber)
+        helixSource: $(_HelixSource)
+
+        # REVIEW: not sure why "cli" is part of the names here. Leave it for the ones that already had it,
+        # but don't add it to new ones.
+        ${{ if ne(parameters.readyToRun, true) }}:
```
Comment (trylek) [inline eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml:55]: Hmm, I overlooked that one, sorry. You no longer need the condition here. In fact, the corresponding parameters at the top of the script can be removed too (readyToRun, crossgen2, compositeBuildMode, runInUnloadableContext, runtimeFlavor, runtimeFlavorDisplayName).

File: src/tests/run.proj (hunk 14 above)
Comment (fanyang-mono) [reply src/tests/run.proj:69]: lol, I will turn revert this change before merged. This is for fast development.

File: src/tests/Common/dirs.proj (hunk 17 above)
Comment (fanyang-mono) [reply src/tests/Common/dirs.proj:21]: You are right, it is just for faster local development.

File: eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml (hunk 18 above)
Comment (fanyang-mono) [reply eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml]: runtimeFlavorDisplayName is still being referenced. I deleted the rest of them.

File: src/tests/Directory.Build.targets (hunk 13 above)
Comment (fanyang-mono) [reply src/tests/Directory.Build.targets:164]: Fixed.

File: src/mono/netcore/sample/Android/Program.csproj [hunk 19]
```diff
@@ -7,8 +7,6 @@
     <MicrosoftNetCoreAppRuntimePackDir>$(ArtifactsBinDir)microsoft.netcore.app.runtime.android-$(TargetArchitecture)\$(Configuration)\runtimes\android-$(TargetArchitecture)</MicrosoftNetCoreAppRuntimePackDir>
     <AndroidAppBuilderDir>$(ArtifactsBinDir)AndroidAppBuilder\$(Configuration)\$(NetCoreAppCurrent)</AndroidAppBuilderDir>
     <EnableTargetingPackDownload>false</EnableTargetingPackDownload>
-    <PublishTrimmed>true</PublishTrimmed>
```
Comment (naricc) [inline src/mono/netcore/sample/Android/Program.csproj:10]: Should changes to the sample project be part of this PR?

File: src/coreclr/tests/runtest.py (hunk 15 above)
Comment (naricc) [reply src/coreclr/tests/runtest.py:1005]: Right now os="Browser" and arch="wasm" are logically equivalent, and will be that way for the foreseeable future. We could make both checks look at the host_os (just check for "Browser").

File: src/mono/netcore/sample/Android/Program.csproj (hunk 19 above)
Comment (fanyang-mono) [reply src/mono/netcore/sample/Android/Program.csproj:10]: That is another PR, which was mentioned in my very first comment of this PR. This change here is a work around I took to make AndroidAppBuilder free from `publish`.

File: eng/pipelines/runtime.yml [hunk 20]
```diff
@@ -415,630 +453,630 @@ jobs:
           eq(dependencies.checkout.outputs['SetPathVars_mono.containsChange'], true),
           eq(variables['isFullMatrix'], true))
       # extra steps, run tests
-      extraStepsTemplate: /eng/pipelines/common/templates/runtimes/wasm-runtime-and-send-to-helix.yml
+      extraStepsTemplate: /eng/pipelines/common/templates/runtimes/android-runtime-and-send-to-helix.yml
```
Comment (safern) [inline eng/pipelines/runtime.yml:456]: instead of introducing an extra step template, use a shared template for both wasm and android, maybe name it `mobile-*` and then based on the `osGroup` parameter, condition the steps and parameters accordingly.

File: src/tests/Common/CLRTest.Execute.Bash.targets (hunk 7 above)
Comment (fanyang-mono) [reply src/tests/Common/CLRTest.Execute.Bash.targets]: I tried to use xharness. And there are compatible issues with runtime tests. One issue is that runtime tests don't generated testResults.xml files. Workarounds needs to be implemented for runtime tests, in order to use xharness on runtime tests.

File: src/coreclr/tests/helixpublishwitharcade.proj (hunk 6 above)
Comment (fanyang-mono) [reply src/coreclr/tests/helixpublishwitharcade.proj]: Fixed, thanks!

Comment (fanyang-mono) [reply src/coreclr/tests/helixpublishwitharcade.proj]: Not sure if we plan to support other arch for brewer, other than wasm. So I didn't touch the logic for wasm.

File: src/tests/Common/android-test-runner/AndroidTestRunner.proj (hunk 16 above)
Comment (fanyang-mono) [reply src/tests/Common/android-test-runner/AndroidTestRunner.proj:16]: We plan to support all of arm64, arm and x64. I will update all places to use arch variable.

Comment (fanyang-mono) [reply src/tests/Common/android-test-runner/AndroidTestRunner.proj:16]: This PR is to support x64 only.

File: src/coreclr/tests/runtest.py (hunk 15 above)
Comment (fanyang-mono) [reply src/coreclr/tests/runtest.py]: Fixed, thanks!

File: eng/pipelines/runtime.yml (hunk 20 above)
Comment (fanyang-mono) [reply eng/pipelines/runtime.yml:456]: Will clean it up in a follow-up PR.

File: src/tests/run.proj [hunk 21]
```diff
@@ -412,6 +470,10 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
              Properties=""
              Condition=" '$(MonoAotCompile)'=='true' " />
 
+    <MSBuild Projects="$(MSBuildProjectFile)"
+             Targets="BuildAndroidApp"
+             Condition=" '$(RunWithAndroid)'=='true' " />
```
Comment (trylek) [inline src/tests/run.proj:475]: I believe that the below clause <code>_CMDDIR=%(TestDirectories.Identity)</code> is what makes MSBuild trigger the build multiple times, once for each item in the <code>TestDirectories</code> item group. Could you try to put a similar clause on the BuildAndroidApp target invocation to see whether it helps? Alternatively the %() clause can be used directly in the BuildAndroidApp target definition similar to what happens here:

https://github.com/dotnet/runtime/blob/b75f8d9fd1b3bcac5d82469bf39f1aea2c3a1652/src/tests/Common/helixpublishwitharcade.proj#L132

Comment (fanyang-mono) [reply src/tests/run.proj:481]: Thanks for the suggestion. The issue has been resolved.

File: src/tests/issues.targets [hunk 22]
```diff
@@ -2626,7 +2626,7 @@
 
 
     <ItemGroup Condition=" '$(TargetArchitecture)' == 'wasm' " >
-   	<ExcludeList Include = "$(XunitTestBinBase)/baseservices/threading/paramthreadstart/ThreadStartBool/**">
+   	    <ExcludeList Include = "$(XunitTestBinBase)/baseservices/threading/paramthreadstart/ThreadStartBool/**">
```
Comment (naricc) [inline src/tests/issues.targets:2629]: Extra whitespace?

Comment (fanyang-mono) [reply src/tests/issues.targets]: Fixed, thanks!

**Comment by fanyang-mono** [comment]
Tests are not able to find Android emulators to run with on helix machine, due to a helix rollout yesterday. The change will be rolled back by https://github.com/dotnet/core-eng/issues/11257

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs [hunk 23]
```diff
@@ -319,6 +319,7 @@ public int RunTest(string executable, string outputFile, string errorFile)
                 process.StartInfo.UseShellExecute = false;
                 process.StartInfo.RedirectStandardOutput = true;
                 process.StartInfo.RedirectStandardError = true;
+                process.StartInfo.EnvironmentVariables.Add("NameSpace", nameSpace);
```
Comment (akoeplinger) [inline src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:322]: this should use a constant similar to `COLLECT_DUMPS_ENVIRONMENT_VAR` instead of hardcoding the name here.

and I think I'd call it `__Namespace` to make it consistent with the other vars. Or maybe it should be `__TestGroup` ?

File: src/tests/run.proj [hunk 24]
```diff
@@ -384,6 +397,75 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
     <MSBuild Projects="$(MSBuildProjectFile)" Targets="CreateXunitWrapper;BuildXunitWrapper" Properties="_CMDDIR=%(TestDirectories.Identity)" />
   </Target>
 
+  <UsingTask TaskName="AndroidAppBuilderTask" AssemblyFile="$(AndroidAppBuilderTasksAssemblyPath)"/>
```
Comment (akoeplinger) [inline src/tests/run.proj:400]: this needs to use a `Condition=" '$(RunWithAndroid)'=='true' "` so we only use it on android

File: src/tests/run.proj [hunk 25]
```diff
@@ -337,6 +340,16 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
                         output.WriteLine(line)%3B
                     }
 
+                    // Add Android app running log to testOutput
+                    if ("$(TargetOS)" == "Android")
+                    {
+                        string androidLogFile = System.IO.Path.GetFullPath(_Global.coreRoot + "/AndroidApps" + "/" + _Global.nameSpace + "/adb-logcat-net.dot." + _Global.nameSpace + ".log")%3B
+                        if(File.Exists(androidLogFile))
+                        {
+                            testOutput.AddRange(System.IO.File.ReadAllLines(androidLogFile))%3B
```
Comment (akoeplinger) [inline src/tests/run.proj:349]: I think we should only add the android log if the test failes, it's not really useful for a successful run and just bloats the results xml.

You can extract the `ret == CoreclrTestWrapperLib.EXIT_SUCCESS_CODE` a few lines below into a `success` variable and only add the log file if it's false.

File: src/tests/run.proj [hunk 26]
```diff
@@ -384,6 +397,75 @@ namespace $([System.String]::Copy($(Category)).Replace(".","_").Replace("\","").
...
   </Target>
 
+  <UsingTask TaskName="AndroidAppBuilderTask" AssemblyFile="$(AndroidAppBuilderTasksAssemblyPath)"/>
+
+  <Target Name="BuildAndroidApp">
+
+    <PropertyGroup>
+      <RuntimeIdentifier>android-$(TargetArchitecture)</RuntimeIdentifier>
+      <CMDDIR_Parent>$([System.IO.Path]::GetDirectoryName($(_CMDDIR)))</CMDDIR_Parent>
+      <CMDDIR_Grandparent>$([System.IO.Path]::GetDirectoryName($(CMDDIR_Parent)))</CMDDIR_Grandparent>
```
Comment (akoeplinger) [inline src/tests/run.proj:407]: I think you can collapse these to:

```suggestion
      <CMDDIR_Grandparent>$([System.IO.Path]::GetDirectoryName($([System.IO.Path]::GetDirectoryName($(_CMDDIR)))))</CMDDIR_Grandparent>
```

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs (hunk 23 above)
Comment (fanyang-mono) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:322]: Changed to `__Namespace`

File: src/tests/run.proj (hunk 24 above)
Comment (fanyang-mono) [reply src/tests/run.proj:400]: Fixed

File: src/tests/run.proj (hunk 25 above)
Comment (fanyang-mono) [reply src/tests/run.proj:349]: Fixed it. However, ideally, I think the log should be handled by xharness.

File: eng/pipelines/common/templates/runtimes/wasm-runtime-and-send-to-helix.yml [hunk 27]
```diff
@@ -53,8 +53,6 @@ steps:
 
         # REVIEW: not sure why "cli" is part of the names here. Leave it for the ones that already had it,
         # but don't add it to new ones.
```
[the remark this answers is not in the document]
Comment (fanyang-mono) [reply eng/pipelines/common/templates/runtimes/wasm-runtime-and-send-to-helix.yml:55]: removed

**Comment by fanyang-mono** [comment]
Tests are randomly failing on Android with the syndrome captured here:
https://github.com/dotnet/xharness/issues/338

**Comment by fanyang-mono** [comment]
Could someone approve this PR please? 
cc: @trylek @akoeplinger @naricc

**Review by naricc — APPROVED** [review]
Please ensure an issue is created for the template clean up @safern  mentioned.

**Review by trylek — APPROVED** [review]
Great work, thank you!

**Comment by jkoritzinsky** [comment]
@fanyang-mono @akoeplinger I'm seeing lots of failures in CI in various Android x64 tests. Can you take a look and disable these tests?

Here's the runfo query that shows all of the failing tests in Android x64 in rolling builds: https://runfo.azurewebsites.net/search/tests/?bq=definition%3Aruntime++started%3A%7E3+kind%3Arolling&tq=jobName%3AAndroid

**Comment by akoeplinger** [comment]
There are multiple issues/causes in that list, one is https://github.com/dotnet/runtime/issues/43927 and another one is https://github.com/dotnet/xharness/issues/338 which hits randomly unfortunately so we can't just disable specific tests.

I think I have an idea how to fix the latter one though :)

**Comment by jkoritzinsky** [comment]
I've opened up an issue to track the failures in dotnet/runtime https://github.com/dotnet/runtime/issues/43983. I'll link it to the xharness issue.
